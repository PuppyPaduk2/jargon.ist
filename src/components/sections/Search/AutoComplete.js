// @flow

import React, { PureComponent } from 'react';
import { navigateTo } from 'gatsby-link';
import { connectAutoComplete } from 'react-instantsearch/connectors';
import Autosuggest from 'react-autosuggest';

import { type Jargon } from '../../../types/jargon';
import { type Tag } from '../../../types/tag';

type IndexType = 'jargons' | 'tags';
type Hit = (Jargon | Tag) & { index: IndexType };

type Section = {
  index: IndexType,
  hits: Array<Jargon | Tag>,
};

type Props = {
  hits: Array<Jargon | Tag>,
  currentRefinement: string,
  refine: string => void,
};

export class AutoCompleteWrapper extends PureComponent<Props> {
  titles: { [string]: string };

  constructor(props: Props) {
    super(props);

    this.titles = {
      jargons: 'Jargonlar',
      tags: 'Konular',
    };
  }

  handleChange = (event: any, { newValue, method }: { newValue: string, method: string }) => {
    if (['click', 'enter'].indexOf(method) > -1) {
      navigateTo(newValue);
    }
  };

  handleSuggestionsFetchRequested = ({ value }: any) => {
    this.props.refine(value);
  };

  handleSuggestionsClearRequested = () => {
    this.props.refine('');
  };

  getSuggestionValue = (hit: Hit) => {
    const prefix = hit.index === 'jargons' ? '' : '/k';

    return `${prefix}/${hit.slug}`;
  };

  getSectionSuggestions = (section: Section) =>
    section.hits.map(hit => ({ ...hit, index: section.index }));

  renderSuggestion = (hit: Hit) => hit.title;

  renderSectionTitle = (section: Section) =>
    section.hits.length > 0 && (
      <h4 className="c-search__suggestions-section-title">{this.titles[section.index]}</h4>
    );

  render() {
    const { hits, currentRefinement } = this.props;

    return (
      <Autosuggest
        suggestions={hits}
        multiSection
        onSuggestionsFetchRequested={this.handleSuggestionsFetchRequested}
        onSuggestionsClearRequested={this.handleSuggestionsClearRequested}
        getSuggestionValue={this.getSuggestionValue}
        getSectionSuggestions={this.getSectionSuggestions}
        renderSuggestion={this.renderSuggestion}
        inputProps={{
          placeholder: 'Jargon veya konu ara',
          value: currentRefinement,
          onChange: this.handleChange,
        }}
        renderSectionTitle={this.renderSectionTitle}
        theme={{
          input: 'c-search__input',
          suggestionsContainer: 'c-search__suggestions-container',
          suggestionsContainerOpen: 'c-search__suggestions-container--open',
          suggestionsList: 'c-search__suggestions-list',
          suggestion: 'c-search__suggestion',
          suggestionHighlighted: 'c-search__suggestion--highlighted',
        }}
      />
    );
  }
}

export default connectAutoComplete(AutoCompleteWrapper);
