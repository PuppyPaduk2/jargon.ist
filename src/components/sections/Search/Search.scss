@import '../../../styles/abstracts/index';

.c-search {
  position: relative;

  &__input {
    background-color: $color-cinder;
    border: 0;
    border-radius: px-to-rem(3px);
    color: $color-white;
    font-size: $g-font-size-small;
    outline: 0;
    padding: px-to-rem(13px) px-to-rem(24px);
    width: 100%;
  }

  &__suggestions-container {
    border-radius: 4px;
    color: $secondary-color;
    position: absolute;
    right: 0;
    width: px-to-rem(400px);

    &--open::before {
      background-color: transparent;
      border-color: transparent transparent $color-tuatara;
      border-style: solid;
      border-width: 0 px-to-rem(10px) px-to-rem(10px);
      content: '';
      display: inline-block;
      margin-left: px-to-rem(365px);
      margin-top: 10px;
      text-align: right;
      vertical-align: middle;
    }
  }

  &__suggestions-section-title {
    background-color: $color-tuatara;
    color: $color-white;
    margin: 0;
    padding: pad(2xsmall) pad(xsmall);
  }

  &__suggestions-list {
    background-color: $color-vapor;
    list-style: none;
    margin: 0;
    padding: 0;
  }

  &__suggestion {
    color: $primary-color;
    padding: pad(2xsmall) pad(xsmall);
  }

  &__suggestion--highlighted {
    background-color: $primary-color;
    color: $color-white;
    cursor: pointer;
  }
}
