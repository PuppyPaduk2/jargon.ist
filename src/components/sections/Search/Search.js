// @flow

import React from 'react';

import { Configure, InstantSearch, Index } from 'react-instantsearch/dom';

import AutoComplete from './AutoComplete';

import './Search.scss';

const Search = () => (
  <div className="c-search">
    <InstantSearch
      appId={process.env.GATSBY_ALGOLIASEARCH_APP_ID}
      apiKey={process.env.GATSBY_ALGOLIASEARCH_SEARCH_KEY}
      indexName="jargons"
    >
      <Configure hitsPerPage={3} />
      <Index indexName="jargons">
        <Configure hitsPerPage={5} />
      </Index>
      <Index indexName="tags">
        <Configure hitsPerPage={5} />
      </Index>
      <AutoComplete />
    </InstantSearch>
  </div>
);

export default Search;
