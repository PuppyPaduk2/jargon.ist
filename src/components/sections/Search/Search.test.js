// @flow

import React from 'react';
import { shallow } from 'enzyme';
import { InstantSearch } from 'react-instantsearch/dom';

import AutoComplete from './AutoComplete';
import Search from './Search';

it('renders correctly', () => {
  const wrapper = shallow(<Search />);

  expect(wrapper).toMatchSnapshot();
});

it('renders Autocomple with InstantSearch', () => {
  const wrapper = shallow(<Search />);

  expect(wrapper.find(AutoComplete)).toHaveLength(1);
  expect(wrapper.find(InstantSearch)).toHaveLength(1);
});
