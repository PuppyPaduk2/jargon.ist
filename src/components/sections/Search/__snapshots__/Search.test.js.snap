// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="c-search"
>
  <CreateInstantSearch
    indexName="jargons"
    refresh={false}
    root={
      Object {
        "Root": "div",
        "props": Object {
          "className": "ais-InstantSearch__root",
        },
      }
    }
  >
    <AlgoliaConfigure(UnknownComponent)
      hitsPerPage={3}
    />
    <CreateIndex
      indexName="jargons"
      root={
        Object {
          "Root": "div",
          "props": Object {
            "className": "ais-MultiIndex__root",
          },
        }
      }
    >
      <AlgoliaConfigure(UnknownComponent)
        hitsPerPage={5}
      />
    </CreateIndex>
    <CreateIndex
      indexName="tags"
      root={
        Object {
          "Root": "div",
          "props": Object {
            "className": "ais-MultiIndex__root",
          },
        }
      }
    >
      <AlgoliaConfigure(UnknownComponent)
        hitsPerPage={5}
      />
    </CreateIndex>
    <AlgoliaAutoComplete(AutoCompleteWrapper) />
  </CreateInstantSearch>
</div>
`;
